lib/health_cards/health_card.rb

```diff
@@ -146,78 +169,90 @@ def to_json(*_args)
         new_url = "resource:#{resource_count}"
         url_map[old_url] = new_url
-        entry['fullUrl'] = new_url
+        entry.fullUrl = new_url
         resource_count += 1
       end
-      [entries, url_map]
+      url_map
     end
 
-    def update_elements(entries)
-      entries.each do |entry|
-        resource = entry['resource']
+    def walk_resource(resource, &block)
+      resource.class::METADATA.each do |field_name, meta|
+        type = meta['type']
+        local_name = meta.fetch :local_name, field_name
+        values = [resource.instance_variable_get("@#{local_name}")].flatten.compact
+        next if values.empty?
 
-        resource.delete('id')
-        resource.delete('text')
-        if resource.dig('meta', 'security')
-          resource['meta'] = resource['meta'].slice('security')
-        else
-          resource.delete('meta')
+        values.each do |value|
+          yield value, type
+          walk_resource value, &block unless FHIR::PRIMITIVES.include? type
         end
-        handle_allowable(resource)
-        update_nested_elements(resource)
       end
     end
 
     def handle_allowable(resource)
-      allowable = self.class.allowable[resource['resourceType']]
-      return resource unless allowable
+      klass = resource.class
+      allowable = self.class.allowable[klass]
 
-      allow = allowable + ['resourceType']
-      resource.select! { |att| allow.include?(att) }
-    end
+      return unless allowable
 
-    def process_url(url)
-      new_url = @url_map.key?(url) ? @url_map[url] : @url_map["#{issuer}/#{url}"]
-      raise InvalidBundleReferenceException, url unless new_url
+      allowed = resource.to_hash.select! { |att| allowable.include?(att) }
 
-      new_url
+      resource.from_hash(allowed)
     end
 
-    def update_nested_elements(hash)
-      hash.map { |k, v| update_nested_element(hash, k, v) }
+    def handle_disallowable(resource)
+      klass = resource.class
+      disallowable = self.class.disallowable[klass]
+
+      return unless disallowable
+
+      allowed = resource.to_hash.delete_if { |att| disallowable.include?(att) }
+
+      resource.from_hash(allowed)
     end
 
-    def update_nested_element(hash, attribute_name, value) # rubocop:disable Metrics/CyclomaticComplexity
-      case value
-      when String
-        hash[attribute_name] = process_url(value) if attribute_name == 'reference'
-      when Hash
-        value.delete('text') if attribute_name.include?('CodeableConcept') || value.key?('coding')
-        update_nested_elements(value)
-      when Array
-        value.each do |member_element|
-          member_element.delete('display') if attribute_name == 'coding'
-          update_nested_elements(member_element) if member_element.is_a?(Hash)
-        end
-      end
+    def process_reference(url_map, ref)
+      url = ref.reference
+
+      return unless url
+
+      new_url = url_map.key?(url) ? url_map[url] : url_map["#{issuer}/#{url}"]
```

Comment:
When is the url map going to contain something of the format "issuer/url"?

Reply:
This allows us to match FHIR references such as "Patient/1" (which is how we create them in our FHIR code). Issuer here is a url so it would be something like "http://example.com/Patient/1". That actually came from a previous PR, it just got moved here.

Reply:
I'm not sure this is quite right since the issuer (which is really just an indication of the endpoint hosting the JWKS) might not be the same as the FHIR endpoint where the FHIR data came from.

If we look at the [FHIR docs for resolving references in a bundle](https://www.hl7.org/fhir/bundle.html#references) we really should "extract the [root] from the fullUrl, and append the reference (type/id) to it". In this case the fullUrI is from the resource containing the reference.